MIG-119 - Backup/Restore progress reporting

pranavgaikwad opened the pull request:
Fixes #685 
Fixes #687 
Fixes #688 

This PR:
* Adds Velero _Backup_ progress to _MigMigration_ CR's `pipeline` field
* Adds Velero _PodVolumeBackup_ progress to _MigMigration_ CR's `pipeline` field
* Reports failed _PodVolumeBackup_ resources whenever a _Backup_ fails completely or partially

```yml
  pipeline:
  - completed: "2020-11-03T14:44:13Z"
    message: Completed
    name: Prepare
    started: "2020-11-03T14:42:23Z"
  - message: Not started
    name: Backup
  - completed: "2020-11-03T14:47:26Z"
    message: Completed
    name: StageBackup
    progress:
    - 'Backup openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-cw4m2:
      Completed'
    - 'PodVolumeBackup openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-cw4m2-8kd89:
      Completed, 1.05 MB out of 1.05 MB backed up (3s)'
    - 'PodVolumeBackup openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-cw4m2-mz88f:
      Completed, 1.05 MB out of 1.05 MB backed up (2s)'
    started: "2020-11-03T14:44:13Z"
  - completed: "2020-11-03T14:48:26Z"
    message: Completed
    name: StageRestore
    progress:
    - 'Restore openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-lklmp:
      Completed'
    - 'PodVolumeRestore openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-lklmp-7vdtr:
      Completed, 1.05 MB out of 1.05 MB restored (6s)'
    - 'PodVolumeRestore openshift-migration/ocp-34824-decrease-data-mig-1604414509-stage-phase-lklmp-xx229:
      Completed, 1.05 MB out of 1.05 MB restored (7s)'
    started: "2020-11-03T14:47:26Z"
  - message: Not started
    name: Restore
  - completed: "2020-11-03T14:49:45Z"
    message: Completed
    name: Final
    started: "2020-11-03T14:48:26Z"
  startTimestamp: "2020-11-03T14:42:23Z"
```

## Parsing in UI

Progress messages follow the following regex to be parsed by UI:

```js
// Match the entire progress message into TWO groups
// Group 1 : Resource 
// Group 2 : Message
const progressLineRe = /([A-Za-z]+ [A-Za-z0-9-]+\/[A-Za-z0-9-]+)*\:* *([A-Za-z0-9 \/\(\)-\.]*)/;

// Match the Resource group into three groups
// Group 1 : Kind
// Group 2 : Namespace
// Group 3 : Name
const resourceRe = /([A-Za-z]+) ([A-Za-z0-9-]+)\/([A-Za-z0-9-]+)/

// Match Group 2 further into 2 groups
// Group 1 : Message
// Group 2 : Duration
const durationRe = /(.*)\(([A-Za-z0-9]+)\)$/
```

## QE Test Results

Image: quay.io/pranavgaikwad/mig-controller:br-progress-11-02

```
EXECUTION RESULT: OK (smoke/ocp-29946-hooksfailure.yml)
EXECUTION RESULT: OK (smoke/ocp-25986-maxpods.yml)
EXECUTION RESULT: OK (smoke/ocp-29918-hooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24871-datagrid.yml)
EXECUTION RESULT: OK (smoke/ocp-24659-mysql.yml)
EXECUTION RESULT: OK (smoke/ocp-30240-datavalidation.yml)
EXECUTION RESULT: OK (smoke/ocp-25021-cronjob.yml)
EXECUTION RESULT: OK (smoke/ocp-32077-empty-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-24995-role.yml)
EXECUTION RESULT: OK (smoke/ocp-31358-restartrestic-newer-than-310.yml)
EXECUTION RESULT: OK (smoke/ocp-26160-max-pvs.yml)
EXECUTION RESULT: OK (smoke/ocp-24686-project.yml)
EXECUTION RESULT: OK (smoke/ocp-25022-cronjob-quiesced.yml)
EXECUTION RESULT: OK (smoke/ocp-24706-basicvolmig.yml)
EXECUTION RESULT: OK (smoke/ocp-24797-mongodb.yml)
EXECUTION RESULT: OK (smoke/ocp-29073-internalis.yml)
EXECUTION RESULT: OK (smoke/ocp-29947-customhooks.yml)
EXECUTION RESULT: OK (smoke/ocp-24787-redis.yml)
EXECUTION RESULT: OK (smoke/ocp-28967-migplan-naming.yml)
EXECUTION RESULT: OK (smoke/ocp-24730-django.yml)
EXECUTION RESULT: OK (smoke/ocp-31309-quotanoattach.yml)
EXECUTION RESULT: OK (smoke/ocp-33378-max-pvs-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-26032-maxns.yml)
EXECUTION RESULT: OK (smoke/ocp-25090-jobs.yml)
EXECUTION RESULT: OK (smoke/ocp-32189-cronjob-with-pvc.yml)
EXECUTION RESULT: OK (smoke/ocp-29871-resquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33393-bz1845569.yml)
EXECUTION RESULT: OK (smoke/ocp-33423-hooksnosa.yml)
EXECUTION RESULT: OK (smoke/ocp-32834-pvc-terminating.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-25000-sets.yml)
EXECUTION RESULT: OK (smoke/ocp-31339-pvc-not-connected.yml)
EXECUTION RESULT: OK (smoke/ocp-25212-initcont.yml)
EXECUTION RESULT: OK (smoke/ocp-33375-maxpods-equalquota.yml)
EXECUTION RESULT: OK (smoke/ocp-33379-maxns-equalquota.yml)
EXECUTION RESULT: FAILURE (smoke/ocp-24872-mssql.yml)
```

djwhatle commented on pkg/controller/migmigration/backup.go:
```diff
@@ -117,6 +117,27 @@ func (t *Task) getStageBackup() (*velero.Backup, error) {
 }
 
+func (t *Task) getPodVolumeBackupsForBackup(backup *velero.Backup) *velero.PodVolumeBackupList {
+	nl := map[string]string{
```
What does `nl` stand for?

pranavgaikwad replied:
@djwhatle definitely needs a better name there. Its the pvb label Velero uses to associate a PVB with a Backup resource

djwhatle commented on pkg/controller/migmigration/backup.go:
```diff
@@ -348,3 +455,18 @@ func findPVVerify(pvList migapi.PersistentVolumes, pvName string) bool {
 	return false
 }
+
+func bytesToSI(bytes int64) string {
```
This is cool, does this convert to whatever the appropriate unit is?

The only downside I see to this is that the user may lose the ability to determine if a PVB is still moving if this auto-switches to GB or anything larger.

djwhatle replied:
Recommend a comment on this function explaining what it does

pranavgaikwad replied:
@djwhatle This comes from the similar logic I used for pvc-migrate. I have not tested this for more than 1 GB. I should. But the formatted string `.2f` should updated the number with a precision of 2 decimal digits : `1.23 GB of 2 GB`. I need to test it though. 

+1 for comment. 

I will add rate calculation in another PR, currently, I am seeing issues with accuracy of the numbers and still need some time to experiment.

djwhatle replied:
oh cool. yeah if we are showing 2 decimal places that seems sufficient to me. Thanks!

djwhatle reviewed (changes requested):
Looking good! I have a few questions.

djwhatle reviewed (approved):
LGTM, would like to see regression test suite results posted here before merge if possible.

alaypatel07 commented on pkg/controller/migmigration/backup.go:
```diff
@@ -117,6 +117,27 @@ func (t *Task) getStageBackup() (*velero.Backup, error) {
+		&list)
+	if err != nil {
+		log.Trace(err)
+		return nil
```
why not return an empty list instead?

This can panic if the user of this function does not check for nil, `return &list` is more graceful IMHO

pranavgaikwad replied:
+1, for the empty list reference, instead of nil

JaydipGabani replied:
+1, for returning an empty list reference

alaypatel07 commented on pkg/controller/migmigration/backup.go:
```diff
@@ -139,12 +160,86 @@ func (t Task) getBackup(labels map[string]string) (*velero.Backup, error) {
+						pvb.Namespace,
+						pvb.Name))
+			}
+		}
```
I would actually consider sorting the `progress` array before returning in order to reduce flapping due to the different order of `pvbList.Items` returned

pranavgaikwad replied:
@alaypatel07 I have not seen that behaviour. The order remains same, maybe that's a wrong assumption to make. Unsure. Sorting based on ongoing phase would be best I think (InProgress at the top, followed by Completed, Failed and Waiting)

alaypatel07 replied:
@pranavgaikwad yes I think it will take a lot of time to observe that behaviour but I don't think the apiserver guarantees order. We should safeguard ourselves against that wherever we can

alaypatel07 replied:
not too opinionated of what would be the best place, will leave that decision to you :)

JaydipGabani replied:
+1 on sorting at the end of the loop. That way we can make sure that the order of PVB remains unchanged.

JaydipGabani replied:
+1 on sorting at the end of loop based on `pvb.Namespace/pvb.Name` to keep the order same everytime.

pranavgaikwad commented:
I made a stupid mistake. I used the same log file to run another set of QE tests which overwrote the results of previous run. I will run again today with the addressed feedback code, then we can merge this PR.

alaypatel07 reviewed (approved):
ACK
